test/e2e/service.go

```diff
@@ -40,6 +40,10 @@ import (
 	"k8s.io/kubernetes/pkg/util/wait"
 )
 
+// Maximum time a kube-proxy daemon on a node is allowed to not
+// notice a Service update, such as type=NodePort.
+const kubeProxyLagTimeout = 5 * time.Minute
```

Comment:
30s isn't guaranteed, but yeah. In the specify case I debugged I saw a lag of 20s. This is in part kube-proxy inefficiently handling each update of a relist of all services one by one, and in part apiserver lag.